pkg/backend/display/progress.go

```diff
@@ -317,37 +257,42 @@ func ShowProgressEvents(op string, action apitype.UpdateKind, stack tokens.Name,
 
 	opStopwatch := newOpStopwatch()
 
+	renderer := &messageRenderer{
+		opts:                  opts,
+		progressOutput:        progressOutput,
+		printedProgressCache:  make(map[string]Progress),
+		nonInteractiveSpinner: spinner,
+	}
+
 	display := &ProgressDisplay{
 		action:                 action,
 		isPreview:              isPreview,
 		opts:                   opts,
+		renderer:               renderer,
 		stack:                  stack,
 		proj:                   proj,
-		progressOutput:         progressOutput,
 		eventUrnToResourceRow:  make(map[resource.URN]ResourceRow),
 		suffixColumn:           int(statusColumn),
 		suffixesArray:          []string{"", ".", "..", "..."},
 		urnToID:                make(map[resource.URN]string),
 		colorizedToUncolorized: make(map[string]string),
-		printedProgressCache:   make(map[string]Progress),
 		displayOrderCounter:    1,
-		nonInteractiveSpinner:  spinner,
 		opStopwatch:            opStopwatch,
 	}
 
 	// Assume we are not displaying in a terminal by default.
-	display.isTerminal = false
+	renderer.isTerminal = false
 	if stdout == os.Stdout {
 		terminalWidth, terminalHeight, err := terminal.GetSize(int(os.Stdout.Fd()))
 		if err == nil {
 			// If the terminal has a size, use it.
-			display.isTerminal = opts.IsInteractive
-			display.terminalWidth = terminalWidth
-			display.terminalHeight = terminalHeight
+			renderer.isTerminal = opts.IsInteractive
```

Comment:
Is this all needed? Won't the first call to render fetch the size and set everything up anyway?